src/daft-core/src/array/ops/utf8.rs

```diff
@@ -1328,6 +1343,44 @@ impl Utf8Array {
         Ok(result)
     }
 
+    pub fn normalize(&self, opts: Utf8NormalizeOptions) -> DaftResult<Utf8Array> {
+        let whitespace_regex = regex::Regex::new(r"\s+").expect("whitespace regex should compile");
+
+        let arrow_result = self
+            .as_arrow()
+            .iter()
+            .map(|maybe_s| {
+                if let Some(s) = maybe_s {
+                    let mut s = s.to_string();
+
+                    if opts.remove_punct {
+                        s = s.replace(PUNCTUATION, "");
+                    }
+
+                    if opts.lowercase {
+                        s = s.to_lowercase();
+                    }
+
+                    if opts.white_space {
+                        s = whitespace_regex
+                            .replace_all(s.as_str().trim(), " ")
+                            .to_string();
```

Comment:
If you ever find that this step is slow, an alternative that may or may not be faster is [str.split_whitespace](https://doc.rust-lang.org/stable/std/primitive.str.html#method.split_whitespace).

Reply:
I'll just keep it for now, it might also be slow to join the string after the split